A-RYTH-MATIK/BitGarden/BitGarden.ino

```diff
@@ -112,80 +143,93 @@ void loop() {
     }
     // Long button press. Change menu page.
     else if (press == Encoder::PRESS_LONG) {
-        if (selected_page == PAGE_MAIN) {
-            selected_param = PARAM_NONE;
-            temp_seed = packet.GetSeed();
-            selected_page = PAGE_SEED;
-        } else {
-            seed_index = 0;
-            packet.UpdateSeed(temp_seed);
-            selected_page = PAGE_MAIN;
-        }
-        update_display = true;
+        HandleLongPress();
     }
 
-    // Read encoder for a change in direction and update the selected parameter.
-    UpdateParameter(hw.encoder.Rotate());
+    // Read encoder for a change in direction and update the selected page or
+    // parameter.
+    (page_select)
+        ? UpdatePage(hw.encoder.Rotate())
+        : UpdateParameter(hw.encoder.Rotate());
 
     // Render any new UI changes to the OLED display.
     UpdateDisplay();
 }
 
 // Initialize random seed and step length from EEPROM or default values.
 void InitState() {
-    // Read previously stored seed from EEPROM memory. If it is empty, generate a new random seed.
-    uint16_t _seed;
-    EEPROM.get(SEED_ADDR, _seed);
-    if (_seed != 0xFFFF) {
-        packet.SetSeed(_seed);
-    } else {
-        packet.NewRandomSeed();
-    }
-    // Read previously stored step_length from EEPROM memory. Update step_length if the value is non-zero.
-    uint8_t _step_length;
-    EEPROM.get(LENGTH_ADDR, _step_length);
-    if (_step_length != 0 && _step_length != 0xFF) {
-        SetLength(_step_length);
+    // Read previously put Storage struct from EEPROM memory. If it doesn't
+    EEPROM.get(0, state);
+
+    // Check if the data in memory matches expected values.
+    if ((strcmp(state.script, SCRIPT_NAME) != 0) || (state.version != SCRIPT_VER)) {
+        // Set script version identifier values.
+        strcpy(state.script, SCRIPT_NAME);
+        state.version = SCRIPT_VER;
+
+        // Default step length.
+        state.step_length = 16;
+
+        // Use a new arbitrary random seed for state.
+        state.seed = 0x8f26;
+
+        // Default mode.
+        state.mode = Mode::TRIGGER;
+
+        // Provide a even distribution of default probabilities.
+        state.probability[0] = 92;
+        state.probability[1] = 86;
+        state.probability[2] = 64;
+        state.probability[3] = 48;
+        state.probability[4] = 32;
+        state.probability[5] = 18;
     }
 }
 
 // Save state to EEPROM if state has changed.
 void SaveChanges() {
     if (!state_changed) return;
     state_changed = false;
-    EEPROM.put(LENGTH_ADDR, step_length);
-    EEPROM.put(SEED_ADDR, packet.GetSeed());
+    state.seed = packet.GetSeed();
+    uint16_t _prob[OUTPUT_COUNT] = {
+        outputs[0].GetProbInt(),
+        outputs[1].GetProbInt(),
+        outputs[2].GetProbInt(),
+        outputs[3].GetProbInt(),
+        outputs[4].GetProbInt(),
+        outputs[5].GetProbInt(),
+    };
+    memccpy(state.probability, _prob, sizeof(_prob), sizeof(_prob));
```

Comment:
Looks like `memccpy` is a typo and I meant to use `memcpy` and only pass in the size of the array once.